File: metodos.js
const math = require("mathjs/lib/browser/math");
const ndarray = require('ndarray');
var zeros = require('zeros'),
    fill = require('ndarray-fill');
 

class methods{

  static biseccion(fx, a, b, maxNum) {
    console.log("Biseccion: ");
    let cond = 0.0000001;
    let i = 1;
    let xm = 0;
    let xm0 = 0;
    let fa = 0;
    let fm = 0;
    let check = 0;
    let error = 1;

    console.log(
      "| i  |     a         |    xm         |      b        |     f(x)   |    E    |"
    );

    while (error > cond && i < maxNum) {
      //Primera iteracion

      if (i == 1) {
        xm = (a + b) / 2;
        fa = math.evaluate(fx);
        fm = fa(xm);
        fa = fa(a);
      } else {
        //Condicion de a y b
        if (check < 0) {
          b = xm;
        } else {
          a = xm;
        }

        xm0 = xm;
        xm = (a + b) / 2;
        fa = math.evaluate(fx);
        fm = fa(xm);
        fa = fa(a);
        error = math.abs(xm - xm0);
      }
      //Valor de check
      if (fm * fa < 0) {
        check = -1;
      } else {
        check = 1;
      }
      console.log(
        "|  " +
          i +
          " | " +
          a.toFixed(10) +
          "  | " +
          xm.toFixed(10) +
          "  | " +
          b.toFixed(10) +
          "  | " +
          fm.toPrecision(6) +
          "  | " +
          error +
          "  |"
      );
      i = i + 1;
    }
  }

  static reglaFalsa(fx, a, b, maxNum) {
    console.log("Regla Falsa:");
    let cond = 0.0000001;
    let i = 1;
    let error = 1;
    let xm = 0;
    let xm0 = 0;
    let f = 0;
    let fa = 0;
    let fb = 0;
    let fm = 0;

    console.log(
      "| i  |     a         |    xm         |     b         |      f(x)    |    E    |"
    );

    while (error > cond && i < maxNum) {
      if (i == 1) {
        f = math.evaluate(fx);
        fa = f(a);
        fb = f(b);
        xm = (fb * a - fa * b) / (fb - fa);
        fm = f(xm);
      } else {
        if (fa * fm < 0) {
          b = xm;
        } else {
          a = xm;
        }
        xm0 = xm;
        fa = f(a);
        fb = f(b);
        xm = (fb * a - fa * b) / (fb - fa);
        fm = f(xm);
        error = math.abs(xm - xm0);
      }
      console.log(
        "|  " +
          i +
          " | " +
          a.toFixed(10) +
          "  | " +
          xm.toFixed(10) +
          "  | " +
          b.toFixed(10) +
          "  | " +
          fm.toPrecision(6) +
          "  | " +
          error +
          "  |"
      );
      i = i + 1;
    }
  }

  static puntoFijo(fx, gx, x0, maxNum) {
    console.log("Punto Fijo:");
    let cond = 0.0000001;
    let error = 1;
    let i = 0;
    let x = x0;
    let f = 0;
    let g = 0;

    console.log(
      "| i  |          xi          |       g(xi)          |    f(xi)  |     E   |"
    );

    while (error > cond && i < maxNum) {
      if (i == 0) {
        f = math.evaluate(fx);
        f = f(x);
        g = math.evaluate(gx);
        g = g(x);
        console.log(
          "|  " +
            i +
            " | " +
            x.toFixed(16) +
            "  | " +
            g +
            "  | " +
            f.toPrecision(6) +
            "  | " +
            error.toPrecision(6) +
            " |    "
        );
      } else {
        x = g;
        f = math.evaluate(fx);
        f = f(x);
        g = math.evaluate(gx);
        g = g(x);
        error = math.abs(x - x0);
        x0 = x;
        console.log(
          "|  " +
            i +
            " | " +
            x +
            "  | " +
            g +
            "  | " +
            f.toPrecision(6) +
            "  | " +
            error.toPrecision(6) +
            "  |    "
        );
      }

      i = i + 1;
    }
  }

  static secante(x0, x1, fx, maxNum) {
    console.log("Secante:");
    let error = 1;
    let i = 0;
    let cond = 0.0000001;
    let f0 = 0;
    let f1 = 0;
    let x = x0;

    console.log(
      "| iter |         xi          |          f(x)       |          E          |"
    );

    while (error > cond && i < maxNum) {
      if (i == 0) {
        f0 = math.evaluate(fx);
        f0 = f0(x0);
        console.log(
          "|  " +
            i +
            "   | " +
            x.toFixed(16) +
            "  | " +
            f1.toFixed(16) +
            "  | " +
            error.toFixed(16) +
            "  |"
        );
      } else if (i == 1) {
        f1 = math.evaluate(fx);
        f1 = f1(x1);
        console.log(
          "|  " +
            i +
            "   | " +
            x1.toFixed(16) +
            "  | " +
            f1 +
            "  | " +
            error.toFixed(16) +
            "  |"
        );
      } else {
        x = x1;
        x1 = x1 - (f1 * (x1 - x0)) / (f1 - f0);
        x0 = x;
        f0 = f1;
        f1 = math.evaluate(fx);
        f1 = f1(x1);
        error = math.abs(x1 - x0);
        console.log(
          "|  " + i + "   | " + x1 + "  | " + f1 + "  | " + error + "  |"
        );
      }

      i = i + 1;
    }
  }
/** 
  static gaussSimple(Ma, b) {
    const matrixA = math.matrix(Ma);
    const vectorB = b;

    const n = matrixA.size()[0];

    let M;
    math.evaluate(`M = [matrixA, vectorB]`, {
      M,
      matrixA,
      vectorB,
    });

    for (let i = 0; i < n - 1; i++) {
      for (let j = i + 1; j < n; j++) {
        if (math.subset(M, math.index(i, j)) !== 0) {
          math.evaluate(`M[j,i:n+1]=M[j,i:n+1]-((M[j,i]/M[i,i])*M[i,i:n+1])`, {
            M,
            n,
            i,
            j,
          });
        }
      }
    }
  }
  */
  
  static vanderMon(x,y) {
        
  if( x.dimension !== 1 ) {
    throw new TypeError('vandermonde: error: x debe ser un vector de una dimensión');
  }
  let inicio= new Date();

  setTimeout(function(){
    terminarScript(inicio);
  },0);

  var i;
  var M = x.shape[0];
  
  var N = arguments[2] || M;
  var reversed = !! arguments[3];
  var dtype = arguments[4] || 'float64';
  var v = zeros([M,N],dtype);
  if( ! reversed ) {
    fill(v,function(i,j) {
      return Math.pow( x.get(i), j );
    });
  } else {
    fill(v,function(i,j) {
      return Math.pow( x.get(M-1-i), j);
    });
  }
  
  var vres = ndarray(new Float64Array(v.size),[math.sqrt(v.size),math.sqrt(v.size)]);
  var cont = x.size-1;
  var cont2 = 0;
    for(let u = 0;u < v.size; u++){
      if(cont< 0){
        cont = x.size-1;
      }
      
      if(u%x.size == 0 && u!=0){
        
        cont2 = cont2+x.size;
      }
      // console.log(v.data[cont2+cont])
      
      vres.data[u] = v.data[cont2+cont];
      
      cont--;
    }
    
    const textMatrix = math.ones(x.size,x.size);
    var matrixres = ""
    
    for(let p=0;p<vres.size/x.size;p++){
      for(let q=0;q<vres.size/x.size;q++){
        textMatrix.subset(math.index(q,p),vres.get(q,p))
       
       matrixres = matrixres + " " + vres.get(p,q) 
      }
      matrixres = matrixres + "\n"
    }
    
    var coeffRes = "";
    var coeff = (math.multiply(math.inv(textMatrix),y))
    var cont1 = x.size-1;
    for(let ty = 0; ty<coeff._size[0];ty++){
      if(ty == coeff._size[0]-1){
        coeffRes = coeffRes + coeff._data[ty] + "x^" + cont1 
      }
      else{
        coeffRes = coeffRes + coeff._data[ty] + "x^" + cont1+ " + " 
      }
    
    cont1--;
    }
    console.log("La Matriz de vandermonde es:")
    console.log()
    console.log(matrixres)
    console.log()
    console.log("El polinomio de la matriz es : ")
    console.log(coeffRes)
    console.log()
    function terminarScript(fechaInicial){
      let fin=new Date();
      let diferencia = fin-fechaInicial;
      console.log("El tiempo que se demora es: " + diferencia + " milisegundos")
    }
    return vres;
        
        
}
  
    //No funcional
    static difDivid2(Vx,Vy) {
      n= math.size(Vx)
      var X = Vx
    
      D=math.zeros(n[0],n[0])
    
      Y = math.transpose(Vy)
      for(let i = 0; i<n; i++ ){
    
        D._data[0,i][0] = Y[i]
        console.log()
      }
    
      for(let i = 1 ; i<n[0];i++){
    
        var aux0 = D._data[math.range(i-1,n[0]),i-1]
        aux = math.diff(aux0)
    
        aux2 = math.subtract(math.subset(X,math.index(math.range(i,n[0]))), math.subset(X,math.index(math.range(0,n[0]-1))))
    
        // D._data[math.range(i,n[0]),i] = math.divide(aux,math.transpose(aux2))
      }
      console.log(aux)
      console.log(math.transpose(aux2))
    
      Coef = math.diag(D)
      // console.log(D._data)
      // console.log(Coef._data)
    }
 
    /**
  static difDivid(Vx,Vy){
    const n = Vx.length;
    const X = math.resize(math.matrix(Vx),[1,4]);
    const Y = math.matrix(Vy);
    let D = math.zeros(n,n);
    const YT = math.resize(Y,[4,1]);
    
    
    D = math.evaluate(`D[:,0]=YT`, {
        D,
        YT,
      });

    
    let aux0;
    for (let i = 0; i < n - 1; i++) {
      math.evaluate(`aux0 = D[i-1:n,i-1]`, {
        aux0,
        D,
        n,
        i,
      });
      aux = math.eval(math.derivative(aux0))
      math.evaluate(`aux2 = X[i:n] - X[0:n-i]`, {
        aux2,
        X,
        n,
        i,
      });
      math.evaluate(`D[i:n,i] = aux/math.transponse(aux2)`, {
        D,
        aux2,
        n,
        i,
      });
    }
    
  }
  */
  static steffenSen( x0,f,tol,nmax ) {
    let x1 = 0
    let x2 = 0
    var x= x0
    x = eval(f)
    x1 = x
    x2 = eval(f)
    var xn = x0 -math.pow((x1-x0),2)/(x2-2*x1+x0)
    let error =math.abs(x2-xn)
    let i = 1
    while(error>=tol && i < nmax){
            x0 = xn
            x =x0
            x1= eval(f)
            x=x1
            x2=eval(f)
            xn = x0 -math.pow((x1-x0),2)/(x2-2*x1+x0)
            error = math.abs(x0-xn);
            i++;
            console.log("Error en la iteracion " + i + " es de " +error)
            console.log("El valor de xn en la iteracion " + i + " es de " +xn)
            console.log()
    }
    console.log(x + " " + x1 )
  }
}

/** 
var k =([[4,-1,0,3],[1,15.5,3,8],[0,-1.3,-4,1.1],[14,5,-2,30]])
var ksp = math.sparse([[4,-1,0,3],[1,15.5,3,8],[0,-1.3,-4,1.1],[14,5,-2,30]])
math.lusolve(k,[1,1,1,1])
console.log(math.lusolve(k,[1,1,1,1]))
*/

// console.log(math.slu(ksp,0,0.0000001))
// console.log(math.lup(k))


//methods.reglaFalsa("f(x)=log(sin(x)^2 + 1) -(0.5)", 0, 1, 100)
//methods.secante(0.5, 1, "f(x) = log(sin(x)^2 + 1) -(0.5)", 100)
//methods.biseccion("f(x) = log(sin(x)^2 + 1) -(0.5)",0,1,100)
//methods.puntoFijo('f(x) = log(sin(x)^2 + 1) - (0.5) -x ', 'f(x) = log(sin(x)^2 + 1) -(0.5)', -0.5, 100)
/** 
var x = ndarray(new Float64Array([-1, 0, 3, 4]));
var h = [15.5, 3, 8, 1];
methods.vanderMon(x,h);
*/

//methods.steffenSen(1.5,'math.sqrt(10/(x+4))',0.00001,10)


//methods.difDivid2([-1,0,3,4],[15.5,3,8,1])

File: metodos.py
import numpy as np
import numpy.matlib
from sympy import sympify, Abs, Symbol
from sympy.core.evalf import evalf


def fixedPoint(fx, gx, x0, tol, numMax):
    results = list()
    results.append(["iter", '{:^15.7}'.format("xi"), '{:^15.7}'.format("g(xi)"), '{:^15.7}'.format("f(xi)"), '{:^15.7}'.format("E")])
    
    x = Symbol('x')
    i = 1
    cond = tol
    error = 1.000
    
    ex = sympify(fx)
    rx = sympify(gx)

    xP = x0
    xA = 0.0

    ea = ex.subs(x, xP)
    ea = ea.evalf()

    ra = rx.subs(x, xP)
    ra = ra.evalf()

    while((error > cond) and (i < numMax)):
    
        ra = rx.subs(x,xP)
        xA = ra.evalf()

        ea = ex.subs(x, xP)
        ea = ea.evalf()

        error = Abs(xA - xP)

        xP = xA

        results.append([i, '{:^15.7f}'.format(float(xA)), '{:^15.7f}'.format(float(ra)), '{:^15.7E}'.format(float(ea)), '{:^15.7E}'.format(float(error))])
        
        i += 1
        
    
    for i in results:
        print(i)

def bisec(a, b, fx, Error,numMax):
    results = list()
    results.append(["iter", '{:^15.7}'.format("a"), '{:^15.7}'.format("xm"), '{:^15.7}'.format("b"), '{:^15.7}'.format("f(xm)"), '{:^15.7}'.format("E")])
        
    x = Symbol('x')
    i = 1
    cond = Error
    error = 1.0000000

    ex = sympify(fx)
    
    ea = ex.subs(x, a)
    ea = ea.evalf()

    xm0 = 0.0
    ex_3 = 0

    xm = (a + b)/2

    ex_3 = ex.subs(x, xm)
    ex_3 = ex_3.evalf()

    while(error > cond) and (i < numMax):
        
        if (ea*ex_3 < 0):
            b = xm
        else:
            a = xm

        xm0 = xm
        xm = (a+b)/2

        ex_3 = ex.subs(x, xm)
        ex_3 = ex_3.evalf()

        error = Abs(xm-xm0)

        results.append([i, '{:^15.7f}'.format(float(a)), '{:^15.7E}'.format(float(xm)), '{:^15.7E}'.format(float(b)), '{:^15.7E}'.format(float(ex_3)), '{:^15.7E}'.format(float(error))])

                        
        i += 1

    for i in results:
        print(i)

def regulaFalsi(a, b, fx, Error,numMax):
    results = list()
    results.append(["iter", '{:^15.7}'.format("a"), '{:^15.7}'.format("xm"), '{:^15.7}'.format("b"), '{:^15.7}'.format("f(xm)"), '{:^15.7}'.format("E")])
    

    x = Symbol('x')
    i = 1
    cond = Error
    error = 1.0000000

    ex = sympify(fx)

    xm = 0
    xm0 = 0
    ex_2 = 0
    ex_3 = 0
    ex_a = 0
    ex_b = 0

    while (error > cond) and (i < numMax):
        if i == 1:
            ex_2 = ex.subs(x, a)
            ex_2 = ex_2.evalf()
            ex_a = ex_2

            ex_2 = ex.subs(x, b)
            ex_2 = ex_2.evalf()
            ex_b = ex_2

            xm = (ex_b*a - ex_a*b)/(ex_b-ex_a)
            ex_3 = ex.subs(x, xm)
            ex_3 = ex_3.evalf()
            results.append([i, '{:^15.7f}'.format(a), '{:^15.7f}'.format(xm), '{:^15.7f}'.format(b), '{:^15.7E}'.format(ex_3)])
        else:

            if (ex_a*ex_3 < 0):
                b = xm
            else:
                a = xm

            xm0 = xm
            ex_2 = ex.subs(x, a)
            ex_2 = ex_2.evalf()
            ex_a = ex_2

            ex_2 = ex.subs(x, b)
            ex_2 = ex_2.evalf()
            ex_b = ex_2

            xm = (ex_b*a - ex_a*b)/(ex_b-ex_a)

            ex_3 = ex.subs(x, xm)
            ex_3 = ex_3.evalf()

            error = Abs(xm-xm0)
            er = sympify(error)
            error = er.evalf()
            results.append([i, '{:^15.7f}'.format(a), '{:^15.7f}'.format(xm), '{:^15.7f}'.format(b), '{:^15.7E}'.format(ex_3), '{:^15.7E}'.format(error)])
        i += 1
    
    for i in results:
        print(i)

def secan(x0, x1, fx, tol, numMax):
    results = list()
    results.append(["iter", '{:^15.7}'.format("xi"), '{:^15.7}'.format("f(xi)"), '{:^15.7}'.format("E")])
    
    x = Symbol('x')
    i = 0
    cond = tol
    error = 1.0000000

    ex = sympify(fx)

    y = x0
    ex_0 = ex
    ex_1 = ex

    while((error > cond) and (i < numMax)):
        if i == 0:
            ex_0 = ex.subs(x, x0)
            ex_0 = ex_0.evalf()
            results.append([i, '{:^15.7f}'.format(float(x0)), '{:^15.7E}'.format(float(ex_0))])
        elif i == 1:
            ex_1 = ex.subs(x, x1)
            ex_1 = ex_1.evalf()
            results.append([i, '{:^15.7f}'.format(float(x1)), '{:^15.7E}'.format(float(ex_1))])
        else:
            y = x1
            x1 = x1 - (ex_1*(x1 - x0)/(ex_1 - ex_0))
            x0 = y

            ex_0 = ex.subs(x, x0)
            ex_0 = ex_1.evalf()

            ex_1 = ex.subs(x, x1)
            ex_1 = ex_1.evalf()

            error = Abs(x1 - x0)
            
            results.append([i, '{:^15.7f}'.format(float(x1)), '{:^15.7E}'.format(float(ex_1)), '{:^15.7E}'.format(float(error))])
        i += 1

    for i in results:
        print(i)

def gaussSimple(Ma, b):
    # Getting matrix dimention
    n = len(Ma)

    # Adding the the vector B at the end of the matrix
    matrixMa = np.matrix(Ma)
    vectorB = np.array(b)
    M = np.column_stack((matrixMa, vectorB))

    steps = {'Step 0': np.copy(M)}

    # Matrix reduction
    for i in range(n-1):
        for j in range(i+1, n):
            if (M[j, i] != 0):
                M[j, i:n+1] = M[j, i:n+1]-(M[j, i]/M[i, i])*M[i, i:n+1]
        steps[f'Step {i+1}'] = np.copy(M)

    x = backSubst(M)

    for i, j in steps.items():
        print(i)
        print(j)
        print("\n")
    print("X")
    print(x)

def vanderMon(Vx,Vy):
    output = list()
    X = np.array(Vx)
    s1 = X.size

    Y = np.array(Vy)

    A = np.zeros((s1,s1))

    i = 0
    for i in range(i,s1):
        A[:,i]=X**(s1-i-1)
    
    Coef = (np.linalg.solve(A,Y)).conj().transpose()
    
    output.append("A")
    output.append(A)
    output.append("Coef")
    output.append(Coef)
    
    for i in output:
        print(i)
        print("\n")
    
def difdivid(Vx,Vy):
    output=list()

    X = np.array(Vx)
    n = X.size

    Y = np.array(Vy)

    D = np.zeros((n,n))

    D[:,0]=Y.T
    for i in range(1,n):
        aux0 = D[i-1:n,i-1]
        aux = np.diff(aux0)
        aux2 = X[i:n] - X[0:n-i]
        D[i:n,i] = aux/aux2.T  

    Coef = np.diag(D)
    
    output.append("D")
    output.append(D)
    output.append("Coef")
    output.append(Coef)
    
    for i in output:
        print(i)
        print("\n")

def LUSimple(Ma, b):
    matrixMa = np.array(Ma).T
    n = matrixMa.shape[0]
    L = np.eye(n)
    U = np.zeros((n,n))
    M = matrixMa
    
    steps = {'Step 0': [np.copy(M)]}
    
    for i in range(n-1):
        for j in range(i+1, n):
            if not (M[j,i] == 0):
                L[j,i]=M[j,i]/M[i,i]
                M[j,i:n]=M[j,i:n]-(M[j,i]/M[i,i])*M[i,i:n]
        U[i, i:n]=M[i,i:n]
        U[i+1,i+1:n]=M[i+1,i+1:n]
        steps[f"Step {i+1}"] = np.copy(M)
        steps[f"Step {i+1}.1"] = {"L:":np.copy(L)}
        steps[f"Step {i+1}.2"] = {"U:":np.copy(U)}
    U[n-1,n-1]=M[n-1,n-1]
    
    z=forSubst(np.column_stack((L,b)))
    x=backSubst(np.column_stack((U,z)))

    for i, j in steps.items():
        print(i)
        print(j)
        print("\n")
    print("X")
    print(x)

def backSubst(M):
    n = M.shape[0]
    x = np.matlib.zeros((n, 1), dtype=complex)
    x[n-1] = M[n-1, n]/M[n-1, n-1]
    for i in range(n-2, -1, -1):
        aux1 = np.hstack((1, np.asarray(x[i+1:n]).reshape(-1)))
        aux2 = np.hstack((M[i, n], np.asarray(-M[i, i+1:n]).reshape(-1)))
        x[i] = np.dot(aux1, aux2)/M[i, i]
    return x

def forSubst(M):
    n = M.shape[0]
    x = np.matlib.zeros((n, 1), dtype=complex)
    x[0] = M[0, n]/M[0, 0]
    for i in range(1, n, 1):
        aux1 = np.hstack((1, np.asarray(x[0:i]).reshape(-1)))
        aux2 = np.hstack((M[i, n], np.asarray(-M[i, 0:i]).reshape(-1)))
        x[i] = np.dot(aux1, aux2)/M[i, i]
    return x

def steffenSen(x0,f,tol,nmax):
    x = Symbol('x')
    
    ex = sympify(f)
    ex1 = ex.subs(x,x0)
    ex1 = ex1.evalf()
    
    x1 = ex1
    
    ex2 = ex.subs(x,x1)
    ex2 = ex2.evalf()
    
    x2 = ex2 
    
    xn = x0 - pow((x1-x0),2)/(x2-2*x1+x0)
    error = abs(x2-xn)
    i = 1
    
    while(error>=tol and i < nmax):
        x0 = xn
        
        ex = sympify(f)
        ex1 = ex.subs(x,x0)
        ex1 = ex1.evalf()
        
        x1 = ex1
        
        ex2 = ex.subs(x,x1)
        ex2 = ex2.evalf()
        
        x2 = ex2 
        xn = x0 - pow((x1-x0),2)/(x2-2*x1+x0)
        error = abs(x0-xn)
        
        i += 1
        
        print(f"El error en la iteración {i} : " + str(error))
        print(f"El valor de x en la iteración {i} : " + str(xn))








#fixedPoint( "ln(sin(x)**2 + 1) - (0.5) -x ", "ln(sin(x)**2 + 1) -(0.5)", -0.5, 0.0000001, 100)
#bisec(0,1,"ln(sin(x)**2 + 1) -(0.5)",0.0000001,100)
#regulaFalsi(0, 1, "ln(sin(x)**2 + 1) -(0.5)", 0.0000001, 100)
#secan(0.5, 1, "ln(sin(x)**2 + 1) -(0.5)", 0.0000001, 100)
A = [
     [2, -1, 0, 3],
     [1, 0.5, 3, 8],
     [0, 13, -2, 11],
     [14, 5, -2, 3]
     ]
b = [1,1,1,1]
#gaussSimple(A,b)
#vanderMon([-1, 0, 3, 4],[15.5, 3, 8, 1])
#difdivid([-1, 0, 3, 4],[15.5, 3, 8, 1])
"""LUSimple([
    [4,-1,0,3],
    [1,15.5,3,8],
    [0,-1.3,-4,1.1],
    [14,5,-2,30]
], [1,1,1,1])
"""
#steffenSen(1.5,'(10/(x+4))^(1/2)',0.00001,10)